src/mlpack/tests/string_encoding_test.cpp

```diff
@@ -501,6 +503,645 @@ BOOST_AUTO_TEST_CASE(CharExtractDictionaryEncodingSerialization)
   EncoderType encoder;
   CharExtract tokenizer;
   arma::mat output;
+  encoder.Encode(stringEncodingInput, output, tokenizer);
+
+  EncoderType xmlEncoder, textEncoder, binaryEncoder;
+  arma::mat xmlOutput, textOutput, binaryOutput;
+
+  SerializeObjectAll(encoder, xmlEncoder, textEncoder, binaryEncoder);
+
+  CheckDictionaries(encoder.Dictionary(), xmlEncoder.Dictionary());
+  CheckDictionaries(encoder.Dictionary(), textEncoder.Dictionary());
+  CheckDictionaries(encoder.Dictionary(), binaryEncoder.Dictionary());
+
+  xmlEncoder.Encode(stringEncodingInput, xmlOutput, tokenizer);
+  textEncoder.Encode(stringEncodingInput, textOutput, tokenizer);
+  binaryEncoder.Encode(stringEncodingInput, binaryOutput, tokenizer);
+
+  CheckMatrices(output, xmlOutput, textOutput, binaryOutput);
+}
+
+BOOST_AUTO_TEST_CASE(BagOfWordsEncodingTest)
+{
+  using DictionaryType = StringEncodingDictionary<boost::string_view>;
+
+  arma::mat output;
+  BagOfWordsEncoding<SplitByAnyOf::TokenType> encoder;
+  SplitByAnyOf tokenizer(" ,.");
+
+  encoder.Encode(stringEncodingInput, output, tokenizer);
+  const DictionaryType& dictionary = encoder.Dictionary();
+
+  // Checking that everything is mapped to different numbers
+  std::unordered_map<size_t, size_t> keysCount;
+  for (auto& keyValue : dictionary.Mapping())
+  {
+    keysCount[keyValue.second]++;
+    // Every token should be mapped only once
+    BOOST_REQUIRE_EQUAL(keysCount[keyValue.second], 1);
+  }
+  arma::mat expected = {
+    {  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0,
+       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 },
+    {  0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1,
+       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 },
+    {  1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1 }
```

Comment:
This seems to give an output shape of 3 rows and 45 columns, so this seems to be row-major (`stringEncodingInput` appears to have three sentences / data points).  Is that right?  I would think that we should have 3 columns (points) and 45 rows (dimensions) in the encoded output.

Reply:
Yes it's row major , should we transform it ???

What is time complexity for transforming a matrix ? 

I guess we should merge the changes as in ,and after that I will just redactor a little bit to change the row major to column based , since presently there are 3 pr , which are dependent and that would create a conflict , so the best now is merge as in , and then just a fresh pr where I just change the row major toh column major

Reply:
@lozhnikov how will we handle without padding case in column major form , that means so coulmn will have more row and some will have less

Reply:
Yes, it's row-major. If I am not mistaken scikit learn does the same.

Reply:
@jeffin143 Do we need to handle the column-major case? I suggest to do as in scikit-learn.

Reply:
The issues I guess rcurtin is mentioning here since mlpack treat data point as a column and mlpack is based on column major , this will create a problem

May be what we can do , is let users know that for all string based transformation we are using row major form

Reply:
Right; what we do here should be the opposite of what scikit does, all just because of the fact that Armadillo is column-major.  I imagine handling any padding should be exactly the same as it is now, just transposed.  For instance, if we were to pass this matrix to the `FFN` class, it would currently need to be transposed, as that (and every mlpack learner*) expects a data point to be a column.

* The LARS algorithm actually does most of its operations on individual dimensions, so it turns out to be computationally advantageous to have the data in a row-major form with Armadillo.  So that algorithm in particular also allows data in a row-major form if the boolean `transposeData` option is false.  But it's the only one, and by default it would expect column-major data.

Reply:
I think armadillo is column-major since LAPACK is column-major. That holds for dense matrices. What about sparse matrices? There are many types like CSR, CSC and so on. I'm just curious about it.

I do agree, we should make the output column-major for armadillo matrices. However, I think we should leave the row-major format for vector<vector<>> since
1. it's actually row-major;
2. mlpack algorithms don't accept `vector<vector<>>` (correct me if I'm wrong);
3. The `vector<vector<>>` output type saves a lot of space in the row-major format in ordinary dictionary encoding;
4. I think the only format useful for the tf-idf and the bag of words algorithms is a sparse matrix. Other formats can't work with large datasets.

How do you think?

Reply:
Agreed.  The Armadillo sparse matrices are column-major too; they use a "hybrid" format that's like a `std::unordered_map<>` along with CSC.  The column-major choice was just to match the dense matrices in Armadillo.

For `vector<vector<>>` I agree, the format you've chosen is best; that can't work with Armadillo.  So we should probably leave that as-is.  (And yeah, mlpack algorithms don't work with that format.)

Reply:
@lozhnikov , Can you cherry-pick the following commit
https://github.com/jeffin143/mlpack/commit/1f78afb5c6e32b8b52e55101773fff8a99d4253d 

So I guess this increase the functionality to users, If a user want it to be in column major he can pass Mat as output type and he will get column major, If he want to play with row-major form he can pass vector<vector<>> as output type, and that would give a row-major form for the user to play around